chapter2.adoc

```diff
@@ -134,68 +132,147 @@ The privilege level of the trigger chain is determined by the trigger enabled fo
+[[ssdextcsr]]
+==== Extension of Sdext CSR
 
-[options="header"]
-[cols="30%,70%"]
-.Tselect, tdata1, tdata2, tdata3 CSR access condition in Debug Mode
-|================================================================
-| Register         |  Access condition
-| tselect(0x7a0)   |  mdbgen == 1 \|\| sdedbgalw == 1
-| tdata1(0x7a1)    |  mdbgen == 1 \|\| sdedbgalw == 1
-| tdata2(0x7a2)    |  mdbgen == 1 \|\| sdedbgalw == 1
-| tdata3(0x7a3)    |  mdbgen == 1 \|\| sdedbgalw == 1
-| tinfo(0x7a4)     |  mdbgen == 1 \|\| sdedbgalw == 1
-|================================================================
-
-The fields in mcontrol, mcontrol6, icount, itrigger, etrigger, and tmexttrigger are read/write accessible only when the access conditions are met. When access is disallowed, writes to these fields are ignored, and reads return zero.
+The following CSRs are provided for debugging supervisor domain. They are only read/write in Debug Mode and not accessible by hart.  
 
+.Allocated addresses for supervisor shadow of Debug Mode CSR  
 [options="header"]
-[cols="20%,80%"]
-.Tdata1 fields access condtion against privilege granted to external debugger
-|====================================
-| Field | Access condition
-| m     | mdbgen == 1                           
-| s     | mdbgen == 1 \|\| sdedbgalw == 1    
-| u     | mdbgen == 1 \|\| sdedbgalw == 1   
-| vs    | mdbgen == 1 \|\| sdedbgalw == 1  
-| vu    | mdbgen == 1 \|\| sdedbgalw == 1
-|====================================
-
-=== Other CSR updates
+[cols="25%,25%,50%"]
+|============================================================================================
+| Number   | Name       | Descirption
+| 0xaaa    | sdcsr      | Supervisor debug control and status register.
+| 0xaaa    | sdpc       | Supervisor debug program counter. 
+| 0xaaa    | sdcratch0  | Supervisor debug scratch register 0.
+| 0xaaa    | sdcratch1  | Supervisor debug scratch register 1.
+|============================================================================================
 
-==== Debug Control and Status (dcsr)
+The `sdcsr` register tracks the current debug state of the hart, formatted as shown in <<sdcsr32>> and <<sdcsr64>>. When `mdbgen` is 0 and `sdedbgalw` is set to 1, the `prv` and `v` fields indicate the the privilege level at which the hart was executing before entering Debug Mode. Meanwhile, the `sdpc` is updated with the address of the next instruction to executed upon entry into Debug Mode. When resuming from Debug Mode, the privilege level of the hart is restored to the values in `prv` and `v`, while the hart's PC is updated wth the address in `sdpc`.   
```

Comment:
I would say this differently.

> The `sdcsr`, `sdpc`, `sdscratch0`, and `sdscratch1` registers are accessible only from Debug Mode, and provide supervisor access to `dcsr`, `dpc`, `dscratch0`, and `dscratch1`, respectively.  The `dscratch0` and `dscratch1` registers are optional, and `sdscratch0` and `sdscratch1` are only required if `dscratch0` and `dscratch1` are implemented.
> 
> The `sdcsr` register provides masked access to `dcsr`, to prevent supervisor access to machine state.  When `dcsr` is accessed through `sdcsr`, the following state modifications apply:
> * `mprven` and `ebreakm` are read-only 0
> * `prv[1]` is read-only 0, which prevents prv from being set to M-mode
> * <...>

This way you aren't duplicating fields defined in the Debug spec, which risks the two specs getting out of sync.  In fact, that already happened, you didn't include the new dcsr.cetrig bit.  But it also makes clear that sdcsr is simply a masked alias of dcsr, not a new physical register.  Much the way stselect is an alias (albeit unmasked) of tselect.  So you only have to define how accesses to sdcsr differ from accesses to dcsr, you don't need to name all the fields that are accessed normally (e.g., v, cause, ebreak{s,u,vs,vu}, etc).

Reply:
Hmm, I have to think about that leakage case.  So your worry is that an M-mode external debugger may populate dpc and dscratch*, then later an S-mode external debugger may be able to read those values if sdpc and sdscratch* are just aliases?  So there is no reset between the use of the two debuggers?  I would think a reset would be required for a change to the lifecycle value, but maybe not for an M-mode change to sdedbgalw.  And I guess M-mode can't then clear dpc/dscratch*, since that can only be done in debug mode.  Though won't entering debug mode always overwrite dpc?  So I don't think that one could leak anything.  But I suppose in theory dscratch* could.

Maybe separate sdscratch* regs is reasonable, though it seems inefficient.  We could also say that changes to sdedbgalw reset dscratch*, not sure if that's better though.

Reply:
As for the sdcsr field specification, I guess you can do it either way, but if you plan to list all the unmasked fields then you probably want a non-normative comment describing why the remaining fields were masked.  That way it's clear which fields you have considered, in case future dscr fields are added.

Reply:
> Maybe separate sdscratch* regs is reasonable, though it seems inefficient. We could also say that changes to sdedbgalw reset dscratch*, not sure if that's better though.

I think it makes sense to clear the scratch register whenever `mdbgen` or `sdedbalw` is cleared. Beside the scenario where the M-mode debugger is deprivileged to S-mode, there could also be cases where multiple supervisor domains are debuggable while others are not. Therefore, it would be more secure to  clear the scratch registers during  context switch and requires no separate physical registers anymore. 
@pdonahue-ventana  do you think this will break any debug use cases?

Reply:
Even simpler, why don't we treat dscratch* like we do the dmode=1 triggers, and say that M-mode can write them when mdbgen=0?  That way dscratch* can be saved & restored (or just cleared) by M-mode on supervisor domain context switch.

Reply:
One reason (and perhaps the only reason) that dscratch registers exist is to support ROM-based debug mode implementations such as https://github.com/chipsalliance/rocket-chip/blob/master/scripts/debug_rom/debug_rom.S (as described in appendix A.2 of the debug spec).

The fact that the ROM writes to dscratch CSRs between commands is why the definitions of dscratch0/1 say:
> A debugger must not write to this register unless hartinfo explicitly mentions it (the Debug Module may use this register internally).

and 3.14.3 says:
> The debugger can make no assumptions about the contents of these registers between commands.

Looking at the RocketChip ROM above, at the start of every program buffer execution the external debugger will always see dscratch0=x8 (and I believe that there is no dscratch1).

If you have an implementation like RocketChip then there is no information leakage.  Before handing control to the external debugger it overwrites dscratch0 with x8 (which is a register that is not hidden from the debugger and thus doesn't leak information).  The only concern would be if you have an implementation that provides dscratch registers above and beyond what are necessary for the ROM (if any).

Frankly, I wonder if sdscratch is necessary at all.  Do debuggers really use dscratch?  Can't a supervisor external debugger survive without those registers (which are optional in Sdext and thus the existing debuggers must be written in a way that wouldn't rely on them)?

Reply:
I am considering a possibility for debug ROM with Sdsec, where debug ROM has only S-mode privilege when mdbgen=0&sdedbgalw=1. In this implementation, `sdscratch*`  would be required, so it seems more friendly to have `sdscratch*` if possible. 

> The debugger can make no assumptions about the contents of these registers between commands.

It is a very strong statement. With this in mind, we could potentially mandate that `dscratch*` are always scrubbed whenever entering debug mode.

Reply:
Ensuring that dscratch* is written by the implementation on every debug mode entry (either cleared by HW or updated by debug ROM) sounds good to me.

Reply:
Thank you for the feedback! I will update to have only `dcsr` and `dpc` are exposed to Supervisor mode. And I'll add a non-normative statement clarifying that S-mode debuggers should not use debug scratch registers.

Reply:
I'm sorry, let me walk back my comment for a moment.  A ROM-based implementation means the ROM code performs register/memory accesses, via Abstract Commands, on behalf of the debugger.  And further, it executes in the same mode as, and with the same privilege as, debugger code in the Program Buffer.  So if we need to block debugger access to dscratch* when medbgen=0, then I think we have to block ROM access.  Which I think will happen naturally, since the ROM code simply runs in Debug Mode.

Unless I'm missing something, rather than require a more complicated ROM-based implementation, I think we would be better off defining sdscratch* addresses that provide supervisor access to dscratch*.  We could require that (s)dscratch* be written on every debug mode entry, which should be natural for all debug ROM implementations.  An alternative would be to make nscratch=0, at least when read through sdcsr.  But that wouldn't actually prevent a debugger from using sdscratch*, it would just ensure that a well-behaved debugger doesn't.

Reply:
> So if we need to block debugger access to dscratch* when medbgen=0, then I think we have to block ROM access. 

I don't understand this.  Is your concern that the ROM is not actually read-only and can somehow be modified?

Reply:
No, I'm not worried about the ROM being compromised.  Rather, I'm worried about the ROM not being able to access dscratch* if medbgen=0.

Reply:
Now I get it.  So we either need sdscratch* and the ROM could unconditionally use that (which would work with either medbgen value) or we need to mandate a more complicated mechanism (like having a Debug-ROM mode vs. a Debug-ProgBuf mode).  Having sdscratch* makes sense to me.

Reply:
Yes, I agree that Debug ROM should reach unconditionally to dscratch CSRs. I can extend this list with mhartid (and potentially more). But I would not worry about Debug ROM execution, since it is implementation specific. Debug ROM execution is something hidden and no need to mention in RISC-V Spec.

Reply:
You said dscratch but I'm hoping that you meant sdscratch.

I agree that we don't need to mention the ROM in this spec (at least not in any normative part). I was just thinking through how it could be done so we don't place unnecessary limitations on that type of implementation.

Reply:
I meant dscratch CSRs. I imagine Debug ROM execution is still at god mode (machine mode) independent from medbgen setting. Maybe I did not explicitly mention about it since I consider it as implementation specific. 
I agree about the separation requirement between Debug-ROM mode vs. a Debug-ProgBuf mode in the implementation somehow. Indeed I think this requirement has already exist before this Secure Debug Specification if you want to support memory access with virtual address in program buffer (if Debug Module supports dcsr.mprven).

Reply:
> I imagine Debug ROM execution is still at god mode (machine mode) independent from medbgen setting.

I don't think that's a good assumption.  In Debug ROM implementations that I've seen (the [PULP implementation](https://github.com/pulp-platform/riscv-dbg/blob/master/doc/debug-system.md) used by OpenTitan), the ROM code is just regular software that runs in Debug Mode.  It has to have the same privilege as the debugger, since it's ROM code that implements things like Abstract Commands on behalf of the debugger.  That's why Paul and I are advocating for sdscratch* CSRs.  They aren't needed by debuggers, but they are needed by the ROM.

That said, looking at the [PULP ROM code](https://github.com/pulp-platform/riscv-dbg/blob/master/debug_rom/debug_rom.S), adding sdscratch* doesn't solve everything.  The ROM code would have to be refactored to access the appropriate \*dscratch\* register based on the debugger privilege.  The ROM park loop also reads mhartid, which would also need a supervisor alias (smhartid).  We should probably get input from people working on these implementations.

Reply:
Good point on mhartid.  Because ROMs are implementation-specific, can we assume that anyone doing this type of implementation should put smhartid in the custom CSR space?  Then we don't consume a CSR number for something that is only useful for a purpose that's not really architecturally visible.

Reply:
> Yes, I agree that Debug ROM should reach unconditionally to dscratch CSRs. I can extend this list with mhartid (and potentially more).

I mentioned about mhartid usage above in my previous comments. If we dive into details of the implementation, I should also mention about the fact that: The implementation specific part (Debug ROM+ Memory mapped Abstract command) should always work with physical address for load/stores. But, program buffer execution should work with both physical/virtual address. As summary, Implementation Specific code execution in debug mode should be differentiated from program buffer execution in different ways.

Reply:
> But, program buffer execution should work with both physical/virtual address.

Program buffer execution uses only physical addresses unless dcsr.mprven=1 and mstatus.MPRV=1.

I've looked at some open source implementations and they all seem to assume that mprven is hardwired to 0 because otherwise the ROM's loads and stores wouldn't work if MPRV=1 when you halt.  In order to do VA-based accesses, OpenOCD has code to emulate page table walks (though the last time I looked, it was only partial emulation and it didn't support some of the permission checks, two stages, or PMP).  Supporting mprven=1 is a lot better for software, though it does make the ROM and hardware slightly more complicated.

Reply:
Yes, Rocketchip Debug Architecture does not support dcsr.mprven=1. But, some proprietary implementations supports it with Debug ROM usage. Sorry If I am repeating myself, but the key point is to differentiate between the implementation specific Code execution and Program buffer Execution debug mode and behave differently. Implementation specific Code execution is not normative in the RISC-V Spec. This differentiation is not supported by open-source debug architectures (Rocketchip and Pulp).

Reply:
I agree with what you say.  I just think that we should think about how you might do a ROM-based implementation.  We shouldn't define the architecture in a way that somehow prevents ROM-based implementations.  Here's one way to do it:
- define a custom CSR romcontrol that is only accessible in Debug mode (similar to dcsr/dpc/dscratch)
- software can clear romcontrol[0] but cannot set it.  Hardware can set it.
- romcontrol[0]=1 means that debug mode can access any CSRs that the ROM needs to access (e.g. mhartid, dscratch0) and it ignores MPRV.
- romcontrol[0]=0 means that debug mode cannot access those CSRs and MPRV behaves as defined by mprven.
- upon entry to debug mode, hardware sets romcontrol[0] (so the ROM can access what it needs)
- ROM does a `csrrci x0, romcontrol, 1` right before jumping to the program buffer
- then the program buffer cannot access these things and it also cannot elevate its own permission due to the rule above about writing romcontrol

Of course, this is just a thought exercise and it's not mandatory to implement it like this.  The only important thing is that it's possible to have a ROM-based implementation.

If you agree with the above then I don't think that we need smhartid and we probably don't need sdscratch*.  What do you think?

Reply:
Yes, I agree with your custom CSR based approach to differentiate them. The only thing I can hightlight is the following step should be done after abstract command execution is completed before program buffer (assuming Rocketchip style Debug ROM: Program Buffer located after abstract command for GPR accesses)

> ROM does a csrrci x0, romcontrol, 1 right before jumping to the program buffer

As summary, I agree that we should think about how Debug ROM based implementations can adapt this specification. One of the valid way is your recommended approach based on Custom CSR usage. I believe we should go to this way instead of specifying sdscratch/smhartid and potentially more (since we do not know all potential Debug ROM implementations).

Reply:
I'll just add my agreement too.  Makes me wonder why we ever standardized dscratch* in the first place.

Reply:
> Makes me wonder why we ever standardized dscratch* in the first place.

Similarly, the dret instruction was standardized in the normative part of [Debug 0.13](https://riscv.org/wp-content/uploads/2019/03/riscv-debug-release.pdf).  [I argued in 2020](https://lists.riscv.org/g/tech-debug/message/223) that this was a microarchitecture thing so we [moved dret](https://github.com/riscv/riscv-debug-spec/pull/547) to the appendix in 1.0.

The dscratch* CSRs might also be details from the initial implementation that were incorrectly elevated to architecture status, though nscratch does kind of make them available architecturally.

Reply:
> I agree with what you say. I just think that we should think about how you might do a ROM-based implementation. We shouldn't define the architecture in a way that somehow prevents ROM-based implementations. Here's one way to do it:
> 
> * define a custom CSR romcontrol that is only accessible in Debug mode (similar to dcsr/dpc/dscratch)
> * software can clear romcontrol[0] but cannot set it.  Hardware can set it.
> * romcontrol[0]=1 means that debug mode can access any CSRs that the ROM needs to access (e.g. mhartid, dscratch0) and it ignores MPRV.
> * romcontrol[0]=0 means that debug mode cannot access those CSRs and MPRV behaves as defined by mprven.
> * upon entry to debug mode, hardware sets romcontrol[0] (so the ROM can access what it needs)
> * ROM does a `csrrci x0, romcontrol, 1` right before jumping to the program buffer
> * then the program buffer cannot access these things and it also cannot elevate its own permission due to the rule above about writing romcontrol
> 
> Of course, this is just a thought exercise and it's not mandatory to implement it like this. The only important thing is that it's possible to have a ROM-based implementation.
> 
> If you agree with the above then I don't think that we need smhartid and we probably don't need sdscratch*. What do you think?

One note on this proposal: this would mean that the hart would need to recognise when it leaves the program buffer code to go back to the ROM park loop, and at that point set romcontrol[0] back to 1. Perhaps that could be added to the ebreak behaviour in ROM-based implementations with program buffers?

Reply:
You're right.  I was implicitly considering the ebreak to be a re-entry to debug mode in the fifth bullet.  It's kind of like when you're in M-mode and you execute an ebreak and you trap to the M-mode handler.  Of course, different implementations may do things differently and the only important thing is that the bit gets set whenever you go to the park loop.

Reply:
I think we all agree that this CSR based approach is just thought exercise to understand ROM-based execution is possible. There are also other ways and they are implementation specific. They will not be in the normative part of the spec.